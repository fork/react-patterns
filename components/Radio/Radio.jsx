import PropTypes from 'prop-types';
import React from 'react';

import StyledRadio from './Radio.style';

const Radio = ({
  disabled,
  error,
  errorMessage,
  id,
  label,
  onChange,
  required,
  value,
  options,
  hideLabel,
  ...props
}) => (
  <StyledRadio disabled={disabled} error={error} hideLabel={hideLabel} {...props}>
    <p className="radio__label">
      {label}
      {required && <span className="radio__required">*</span>}
    </p>
    {options.map(option => (
      <div className="radio__item" key={option.value}>
        <input
          id={option.value}
          value={option.value}
          name={id}
          type="radio"
          onChange={onChange}
          disabled={disabled}
          required={required}
          checked={value === option.value}
        />
        <label htmlFor={option.value}>
          <span className={`radio__custom-icon ${value === option.value ? 'checked' : ''}`} />
          {option.label}
        </label>
      </div>
    ))}
    {error && errorMessage && <span className="radio__error">{errorMessage}</span>}
  </StyledRadio>
);

Radio.propTypes = {
  disabled: PropTypes.bool,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  hideLabel: PropTypes.bool,
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  required: PropTypes.bool,
  value: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired
    })
  ).isRequired
};

Radio.defaultProps = {
  disabled: false,
  error: false,
  errorMessage: '',
  hideLabel: false,
  onChange: () => {},
  required: false,
  value: ''
};

export default Radio;
