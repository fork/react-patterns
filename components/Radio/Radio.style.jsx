import styled, { css } from 'styled-components';
import { color, space, hideVisually } from '../../stylesheets';

const StyledRadio = styled.div`
  .radio__label {
    font-weight: bold;
    margin-bottom: ${space('s')};
    ${props => props.hideLabel && hideVisually()}
  }

  .radio__item {
    display: flex;
    align-items: center;
    margin-bottom: ${space('s')};
  }

  label {
    position: relative;
    cursor: pointer;
    display: flex;
    align-items: center;
    color: ${color('black')};
    line-height: 1;
  }

  .radio__custom-icon {
    position: relative;
    display: inline-block;
    width: ${space('l')};
    height: ${space('l')};
    border: 1px solid ${color('black')};
    margin-right: ${space('xs')};
    border-radius: 100%;

    &:after {
      display: none;
      position: absolute;
      content: '';
      border-radius: 100%;
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      width: ${space('s')};
      height: ${space('s')};
      background: ${color('black')};
    }
  }

  input {
    ${hideVisually()}
  }

  input[disabled] ~ label {
    opacity: 0.5;
    cursor: not-allowed;
  }

  input:checked ~ label > .radio__custom-icon:after {
    display: block;
  }

  .radio__wrapper {
    position: relative;
  }

  .radio__error {
    display: block;
    color: ${color('primary')};
    margin-top: ${space('s')};
  }

  ${props =>
    props.error &&
    css`
      .radio__label {
        color: ${color('primary')};
      }

      label .radio__custom-icon {
        border-color: ${color('primary')};
      }
    `}
`;

export default StyledRadio;
