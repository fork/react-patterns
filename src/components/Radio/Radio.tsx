import React from 'react';

import Text from '../Text';

import StyledRadio from './Radio.style';

export type RadioOption = {
  value: string;
  label: string;
};

export type RadioProps = {
  disabled?: boolean;
  error?: boolean;
  errorMessage?: string;
  id: string;
  label: string;
  onChange?: (ev: React.ChangeEvent<HTMLInputElement>) => void;
  required?: boolean;
  success?: boolean;
  value?: string;
  options: RadioOption[];
  hideLabel?: boolean;
};

const Radio = ({
  disabled,
  error,
  errorMessage,
  id,
  label,
  onChange,
  required,
  success,
  value,
  options,
  hideLabel
}: RadioProps) => (
  <StyledRadio error={error} hideLabel={hideLabel} success={success}>
    <p className="radio__label">
      <Text as="span" variant="form-label">
        {label}
        {required && <span className="radio__required">*</span>}
      </Text>
    </p>
    {options.map(option => (
      <div className="radio__item" key={option.value}>
        <input
          id={option.value}
          value={option.value}
          name={id}
          type="radio"
          onChange={onChange}
          disabled={disabled}
          required={required}
          checked={value === option.value}
        />
        <label htmlFor={option.value}>
          <span className={`radio__custom-icon ${value === option.value ? 'checked' : ''}`} />
          <Text as="span" variant="form-value">
            {option.label}
          </Text>
        </label>
      </div>
    ))}

    {error && errorMessage && (
      <small className="radio__error" role="status">
        <Text as="span" variant="copy-small">
          {errorMessage}
        </Text>
      </small>
    )}
  </StyledRadio>
);

export default Radio;
