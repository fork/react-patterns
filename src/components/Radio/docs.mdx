import { Story, Canvas } from '@storybook/addon-docs/blocks';

# Radio

<Canvas>
  <Story id="components-radio--without-value" />
</Canvas>
