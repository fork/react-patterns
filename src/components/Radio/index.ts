export { default, RadioProps, RadioOption } from './Radio';
