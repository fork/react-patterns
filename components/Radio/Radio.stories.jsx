import React, { useState } from 'react';

import Radio from './Radio';

export default {
  title: 'Components/Radio',
  component: Radio
};

const options = [
  {
    value: 'cao',
    label: '🍲 Cao'
  },
  {
    value: 'polo',
    label: '🏇 Polokantine'
  },
  {
    value: 'ottos',
    label: '🍔 Ottos'
  }
];

export const withState = () => {
  const [value, setValue] = useState('frau');

  return (
    <Radio
      id="radio-1"
      label="What do we eat for lunch?"
      value={value}
      options={options}
      onChange={e => setValue(e.target.value)}
    />
  );
};

export const withoutValue = () => (
  <Radio id="radio-2" label="What do we eat for lunch?" options={options} />
);

export const withValue = () => (
  <Radio id="radio-3" label="What do we eat for lunch?" value="ottos" options={options} />
);

export const withoutLabel = () => (
  <Radio id="radio-4" label="What do we eat for lunch?" value="ottos" options={options} hideLabel />
);

export const error = () => (
  <Radio
    id="radio-5"
    label="What do we eat for lunch?"
    options={options}
    error
    errorMessage="We need to eat something. Choose wisely."
    required
  />
);

export const disabled = () => (
  <Radio id="radio-6" label="What do we eat for lunch?" options={options} disabled />
);
