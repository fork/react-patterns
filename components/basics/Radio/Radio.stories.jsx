import React, { useState } from 'react';

import Radio from './Radio';

export default {
  title: 'Basics/Radio',
  component: Radio
};

const options = [
  {
    value: 'frau',
    label: 'Frau'
  },
  {
    value: 'herr',
    label: 'Herr'
  },
  {
    value: 'divers',
    label: 'Divers'
  }
];

export const withState = () => {
  const [value, setValue] = useState('frau');

  return (
    <Radio
      id="radio-1"
      label="Anrede"
      value={value}
      options={options}
      onChange={e => setValue(e.target.value)}
    />
  );
};

export const withoutValue = () => <Radio id="radio-2" label="Anrede" options={options} />;

export const withValue = () => (
  <Radio id="radio-3" label="Anrede" value="divers" options={options} />
);

export const withoutLabel = () => (
  <Radio id="radio-4" label="Anrede" value="divers" options={options} hideLabel />
);

export const error = () => (
  <Radio
    id="radio-5"
    label="Anrede"
    options={options}
    error
    errorMessage="Dies ist ein Pflichtfeld"
    required
  />
);

export const disabled = () => <Radio id="radio-6" label="Anrede" options={options} disabled />;
