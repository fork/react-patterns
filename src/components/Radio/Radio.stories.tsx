import React, { useEffect, useState } from 'react';
import { Story } from '@storybook/react';

import Radio, { RadioProps } from './Radio';
import docs from './docs.mdx';

export default {
  title: 'Components/Input/Radio',
  component: Radio,
  parameters: {
    docs: {
      page: docs
    }
  }
};

const defaults: RadioProps = {
  id: 'radio-id',
  label: 'What do we eat for lunch?',
  required: true,
  error: false,
  disabled: false,
  hideLabel: false,
  errorMessage: 'We need to eat something. Choose wisely.',
  options: [
    {
      value: 'cao',
      label: 'üç≤ Cao'
    },
    {
      value: 'polo',
      label: 'üèá Polokantine'
    },
    {
      value: 'ottos',
      label: 'üçî Ottos'
    }
  ]
};

const Template: Story<RadioProps> = ({ value, ...args }) => {
  const [val, setValue] = useState(value);

  useEffect(() => {
    setValue(value);
  }, [value]);

  return <Radio value={val} onChange={e => setValue(e.target.value)} {...args} />;
};

export const Default = Template.bind({});
Default.args = defaults;

export const WithValue = Template.bind({});
WithValue.args = { ...defaults, value: 'ottos' };

export const WithoutLabel = Template.bind({});
WithoutLabel.args = { ...defaults, value: 'ottos', hideLabel: true };

export const Success = Template.bind({});
Success.args = { ...defaults, value: 'ottos', success: true };

export const Error = Template.bind({});
Error.args = { ...defaults, error: true };

export const Disabled = Template.bind({});
Disabled.args = { ...defaults, disabled: true };
