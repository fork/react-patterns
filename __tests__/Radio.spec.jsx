import React from 'react';
import { shallow } from 'enzyme';

import Radio from '../components/Radio';

const options = [
  {
    value: 'frau',
    label: 'Frau'
  },
  {
    value: 'herr',
    label: 'Herr'
  },
  {
    value: 'divers',
    label: 'Divers'
  }
];

describe('Radio', () => {
  test('Radio should render correctly', () => {
    const component = shallow(<Radio label="Radio Label" options={options} id="1" />);

    expect(component).toMatchSnapshot();
  });

  test('Radio without label should render correctly', () => {
    const component = shallow(<Radio hideLabel label="Radio Label" options={options} id="1" />);

    expect(component).toMatchSnapshot();
  });

  test('Radio with default value should render correctly', () => {
    const component = shallow(
      <Radio hideLabel label="Radio Label" options={options} value="frau" id="1" />
    );

    expect(component).toMatchSnapshot();
  });
});
