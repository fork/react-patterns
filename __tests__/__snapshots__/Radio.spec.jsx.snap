// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Radio should render correctly 1`] = `
<styled.div
  disabled={false}
  error={false}
  hideLabel={false}
>
  <p
    className="radio__label"
  >
    Radio Label
  </p>
  <div
    className="radio__item"
    key="frau"
  >
    <input
      checked={false}
      disabled={false}
      id="frau"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="frau"
    />
    <label
      htmlFor="frau"
    >
      <span
        className="radio__custom-icon "
      />
      Frau
    </label>
  </div>
  <div
    className="radio__item"
    key="herr"
  >
    <input
      checked={false}
      disabled={false}
      id="herr"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="herr"
    />
    <label
      htmlFor="herr"
    >
      <span
        className="radio__custom-icon "
      />
      Herr
    </label>
  </div>
  <div
    className="radio__item"
    key="divers"
  >
    <input
      checked={false}
      disabled={false}
      id="divers"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="divers"
    />
    <label
      htmlFor="divers"
    >
      <span
        className="radio__custom-icon "
      />
      Divers
    </label>
  </div>
</styled.div>
`;

exports[`Radio Radio with default value should render correctly 1`] = `
<styled.div
  disabled={false}
  error={false}
  hideLabel={true}
>
  <p
    className="radio__label"
  >
    Radio Label
  </p>
  <div
    className="radio__item"
    key="frau"
  >
    <input
      checked={true}
      disabled={false}
      id="frau"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="frau"
    />
    <label
      htmlFor="frau"
    >
      <span
        className="radio__custom-icon checked"
      />
      Frau
    </label>
  </div>
  <div
    className="radio__item"
    key="herr"
  >
    <input
      checked={false}
      disabled={false}
      id="herr"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="herr"
    />
    <label
      htmlFor="herr"
    >
      <span
        className="radio__custom-icon "
      />
      Herr
    </label>
  </div>
  <div
    className="radio__item"
    key="divers"
  >
    <input
      checked={false}
      disabled={false}
      id="divers"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="divers"
    />
    <label
      htmlFor="divers"
    >
      <span
        className="radio__custom-icon "
      />
      Divers
    </label>
  </div>
</styled.div>
`;

exports[`Radio Radio without label should render correctly 1`] = `
<styled.div
  disabled={false}
  error={false}
  hideLabel={true}
>
  <p
    className="radio__label"
  >
    Radio Label
  </p>
  <div
    className="radio__item"
    key="frau"
  >
    <input
      checked={false}
      disabled={false}
      id="frau"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="frau"
    />
    <label
      htmlFor="frau"
    >
      <span
        className="radio__custom-icon "
      />
      Frau
    </label>
  </div>
  <div
    className="radio__item"
    key="herr"
  >
    <input
      checked={false}
      disabled={false}
      id="herr"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="herr"
    />
    <label
      htmlFor="herr"
    >
      <span
        className="radio__custom-icon "
      />
      Herr
    </label>
  </div>
  <div
    className="radio__item"
    key="divers"
  >
    <input
      checked={false}
      disabled={false}
      id="divers"
      name="1"
      onChange={[Function]}
      required={false}
      type="radio"
      value="divers"
    />
    <label
      htmlFor="divers"
    >
      <span
        className="radio__custom-icon "
      />
      Divers
    </label>
  </div>
</styled.div>
`;
